shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/OverlaySchedule.hs

```diff
@@ -0,0 +1,191 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Shelley.Spec.Ledger.OverlaySchedule
+  ( -- * Overlay schedule
+    OverlaySchedule,
+    OverlaySlots,
+    compactOverlaySchedule,
+    decompactOverlaySchedule,
+    emptyOverlaySchedule,
+    isOverlaySlot,
+    lookupInOverlaySchedule,
+    overlaySchedule,
+    overlayScheduleHelper,
+    overlayScheduleIsEmpty,
+    overlayScheduleToMap,
+    overlaySlots,
+
+    -- * OBftSlot (for testing)
+    OBftSlot (..),
+  )
+where
+
+import Cardano.Binary
+  ( FromCBOR (..),
+    ToCBOR (..),
+    TokenType (TypeNull),
+    decodeNull,
+    encodeNull,
+    peekTokenType,
+  )
+import Cardano.Prelude (NFData, NoUnexpectedThunks)
+import Cardano.Slotting.Slot
+import Control.Monad.Trans.Reader (asks)
+import Data.List.NonEmpty (NonEmpty)
+import qualified Data.List.NonEmpty as NonEmpty
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Set (Set)
+import qualified Data.Set as Set
+import GHC.Generics (Generic)
+import Shelley.Spec.Ledger.BaseTypes
+import Shelley.Spec.Ledger.Crypto (Crypto)
+import Shelley.Spec.Ledger.Keys
+  ( KeyHash (..),
+    KeyRole (..),
+  )
+import Shelley.Spec.Ledger.PParams (PParams, _d)
+import Shelley.Spec.Ledger.Slot
+
+data OBftSlot crypto
+  = NonActiveSlot
+  | ActiveSlot !(KeyHash 'Genesis crypto)
+  deriving (Show, Eq, Ord, Generic)
+
+instance
+  Crypto crypto =>
+  ToCBOR (OBftSlot crypto)
+  where
+  toCBOR NonActiveSlot = encodeNull
+  toCBOR (ActiveSlot k) = toCBOR k
+
+instance
+  Crypto crypto =>
+  FromCBOR (OBftSlot crypto)
+  where
+  fromCBOR = do
+    peekTokenType >>= \case
+      TypeNull -> do
+        decodeNull
+        pure NonActiveSlot
+      _ -> ActiveSlot <$> fromCBOR
+
+instance NoUnexpectedThunks (OBftSlot crypto)
+
+instance NFData (OBftSlot crypto)
+
+newtype OverlaySchedule crypto = OverlaySchedule (Map SlotNo (OBftSlot crypto))
+  deriving stock (Show, Eq)
+  deriving newtype (NoUnexpectedThunks, NFData)
+
+emptyOverlaySchedule :: OverlaySchedule crypto
+emptyOverlaySchedule = OverlaySchedule Map.empty
```

Comment:
If we intend this API to be qualified when imported, we could have shorter names like `empty`, `lookup`, `isEmpty`, `new/create/...`, `toMap`, `compact`, `decompact`, etc.

Reply:
Yes, considered that but couldn't be bothered to change things around yet...